charms/metallb-speaker/src/charm.py

```diff
@@ -0,0 +1,244 @@
+#!/usr/bin/env python3
+"""Speaker component for the MetalLB bundle."""
+
+import logging
+import os
+from base64 import b64encode
+
+from ops.charm import CharmBase
+from ops.framework import StoredState
+from ops.main import main
+from ops.model import (
+    ActiveStatus,
+    BlockedStatus,
+    MaintenanceStatus,
+)
+
+import utils
+
+logger = logging.getLogger(__name__)
+
+
+class MetallbSpeakerCharm(CharmBase):
+    """MetalLB Speaker Charm."""
+
+    _stored = StoredState()
+
+    def __init__(self, *args):
+        """Charm initialization for events observation."""
+        super().__init__(*args)
+        self.framework.observe(self.on.start, self.on_start)
+        self.framework.observe(self.on.remove, self.on_remove)
+        # -- initialize states --
+        self._stored.set_default(started=False)
+        # -- base values --
+        self._stored.set_default(namespace=os.environ["JUJU_MODEL_NAME"])
+        self._stored.set_default(container_image='metallb/speaker:v0.9.3')
+
+    def on_start(self, event):
+        """Occurs upon start or installation of the charm."""
+        if not self.framework.model.unit.is_leader():
```

Comment:
Doesn't really matter, but the other charm uses the `self.model` shortcut, so maybe use that here, too, for consistency?

Reply:
Sure, I don't really see why there are 2 ways to call that. I think I got the framework.model.unit.is_leader() from the charm template itself. It might be the "old" way and it has been refractored

Reply:
from charmcraft team : CharmBase has app, model, unit, meta, and charm_dir all proxied to the appropriate places

So I'll be removing the extra "framework" at all places that I can

Reply:
Yeah, the `CharmBase` attributes were added later as shortcuts.  It's fine to use either, but I think the shorter version is generally preferred.

Reply:
Update pushed.